subprojects/logging/src/integTest/groovy/org/gradle/internal/logging/ConsoleJvmTestLoggingFunctionalTest.groovy

```diff
@@ -0,0 +1,225 @@
+        executer.withStackTraceChecksDisabled()
+
+        given:
+        buildFile << testLoggingEvents(TestLogEvent.STARTED.testLoggingIdentifier, TestLogEvent.FAILED.testLoggingIdentifier)
+        buildFile << testLoggingStandardStream()
+
+        file(JAVA_TEST_FILE_PATH) << javaTestClass {
+            """
+                System.out.println("standard output");
+                throw new RuntimeException("expected");
+            """
+        }
+
+        when:
+        fails(TEST_TASK_NAME)
+
+        then:
+        def taskOutput = getTaskOutput(result)
+        taskOutput.contains("""MyTest > testExpectation ${TestLogEvent.STANDARD_OUT.consoleMarker}
+    standard output""")
+        matchesTaskOutput(taskOutput, testLogEventRegex(TestLogEvent.STARTED.consoleMarker))
+        matchesTaskOutput(taskOutput, testLogEventRegex(TestLogEvent.FAILED.consoleMarker))
+    }
+
+    def "can group output from custom test listener with task"() {
+        buildFile << """
+            test {
+                beforeTest { descriptor ->
+                    logger.quiet 'Starting test: ' + descriptor.className + ' > ' + descriptor.name
+                }
+                afterTest { descriptor, result ->
+                    logger.quiet 'Finishing test: ' + descriptor.className + ' > ' + descriptor.name
+                }
+            }
+        """
+        file(JAVA_TEST_FILE_PATH) << javaTestClass { '' }
+
+        when:
+        succeeds(TEST_TASK_NAME)
+
+        then:
+        def taskOutput = getTaskOutput(result)
+        taskOutput.contains('Starting test: MyTest > testExpectation')
+        taskOutput.contains('Finishing test: MyTest > testExpectation')
+    }
+
+    static String javaProject() {
+        """
+            apply plugin: 'java'
+            
+            repositories {
+                mavenCentral()
+            }
+            
+            dependencies {
+                testCompile 'junit:junit:4.12'
+            }
+        """
+    }
+
+    static String testLoggingEvents(String... events) {
+        """
+            test {
+                testLogging {
+                    events ${events.collect { "'$it'" }.join(', ')}
+                }
+            }
+        """
+    }
+
+    static String testLoggingStandardStream() {
+        """
+            test {
+                testLogging {
+                    showStandardStreams = true
+                }
+            }
+        """
+    }
+
+    static String javaTestClass(Closure<String> testMethodBody) {
+        """
+            import org.junit.Test;
+
+            public class MyTest {
+                @Test
+                public void testExpectation() {
+                    ${testMethodBody()}
+                }
+            }
+        """
+    }
+
+    static String getTaskOutput(ExecutionResult result) {
+        def taskOutput = result.groupedOutput.task(TEST_TASK_PATH).output
+        assert taskOutput != null
```

Comment:
this should probably be in the fixture.  If the task can't be found, we should produce a nice assertion.  Then you can just inline this.